tools/wake-format/formatter.h

```diff
@@ -0,0 +1,243 @@
+    action1.run(builder, ctx, node);
+    action2.run(builder, ctx, node);
+  }
+};
+
+template <class F>
+struct WalkAction {
+  F walker;
+  uint8_t node_type;
+
+  WalkAction(F walk, uint8_t node_type) : walker(walk), node_type(node_type) {}
+
+  void run(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node) {
+    assert(node.id() == node_type);
+    auto doc = walker(ctx, const_cast<const CSTElement&>(node));
+    builder.append(doc);
+    node.nextSiblingElement();
+  }
+};
+
+template <class F, class P>
+struct WalkPredicateAction {
+  F walker;
+  P predicate;
+
+  WalkPredicateAction(F walk, P p) : walker(walk), predicate(p) {}
+
+  void run(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node) {
+    assert(predicate(node));
+    auto doc = walker(ctx, const_cast<const CSTElement&>(node));
+    builder.append(doc);
+    node.nextSiblingElement();
+  }
+};
+
+// This is the escape hatch to implement something else.
+// NOTE: You're responsible for advancing the node!!!
+template <class F>
+struct EscapeAction {
+  F f;
+  EscapeAction(F f) : f(f) {}
+
+  void run(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node) {
+    // You have to do everything yourself here, that's the price of an escape hatch
+    // we can build up enough of these that it shouldn't be an issue however.
+    f(builder, ctx, node);
+  }
+};
+
+// This does nothing, good for kicking off a chain of formatters
+struct EpsilonAction {
+  void run(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node) {}
+};
+
+template <class Action>
+struct Formatter {
+  Action action;
+  Formatter(Action a) : action(a) {}
+
+  Formatter<SeqAction<Action, ConsumeWhitespaceAction>> consume_wsnl() {
+    return {SeqAction<Action, ConsumeWhitespaceAction>(action, ConsumeWhitespaceAction{})};
+  }
+
+  Formatter<SeqAction<Action, WhitespaceTokenAction>> ws() {
+    return {SeqAction<Action, WhitespaceTokenAction>(action, WhitespaceTokenAction{})};
+  }
+
+  Formatter<SeqAction<Action, SpaceAction>> space(uint8_t count = 1) {
+    return {SeqAction<Action, SpaceAction>(action, SpaceAction{count})};
+  }
+
+  Formatter<SeqAction<Action, NewlineAction>> newline() {
+    return {SeqAction<Action, NewlineAction>(action, NewlineAction{})};
+  }
+
+  Formatter<SeqAction<Action, TokenAction>> token(uint8_t id) {
+    return {SeqAction<Action, TokenAction>(action, TokenAction{id})};
+  }
+
+  Formatter<SeqAction<Action, TokenReplaceAction>> token(uint8_t id, const char* str) {
+    return {SeqAction<Action, TokenReplaceAction>(action, TokenReplaceAction{id, str})};
+  }
+
+  template <class Walker>
+  Formatter<SeqAction<Action, WalkAction<Walker>>> walk(uint8_t node_type, Walker texas_ranger) {
+    return {
+        SeqAction<Action, WalkAction<Walker>>(action, WalkAction<Walker>{texas_ranger, node_type})};
+  }
+
+  template <class Walker>
+  Formatter<
+      SeqAction<Action, WalkPredicateAction<Walker, std::function<bool(const CSTElement& node)>>>>
```

Comment:
These should not be std::functions, define a function object and use that for this case. std::function can't be inlined.